lib/Sema/TypeCheckDeclOverride.cpp

```diff
@@ -591,6 +591,85 @@ static bool parameterTypesMatch(const ValueDecl *derivedDecl,
   return true;
 }
 
+// SWIFT_ENABLE_TENSORFLOW
+static bool overridesDifferentiableAttribute(ValueDecl *derivedDecl,
+                                             ValueDecl *baseDecl) {
+  ASTContext &ctx = derivedDecl->getASTContext();
+  auto &diags = ctx.Diags;
+
+  auto *derivedAFD = dyn_cast<AbstractFunctionDecl>(derivedDecl);
+  auto *baseAFD = dyn_cast<AbstractFunctionDecl>(baseDecl);
+
+  if (!derivedAFD || !baseAFD)
+    return false;
+
+  auto derivedDAs = derivedAFD->getAttrs().getAttributes<DifferentiableAttr>();
+  auto baseDAs = baseAFD->getAttrs().getAttributes<DifferentiableAttr>();
+
+  // Make sure all the differentiable attributes in `baseDecl` are
+  // also declared in `derivedDecl`.
+  for (auto baseDA : baseDAs) {
+    auto baseParameters = AutoDiffIndexSubset::get(
+        ctx, baseDA->getParameterIndices()->parameters);
+    auto defined = false;
+    for (auto derivedDA : derivedDAs) {
+      auto derivedParameters = AutoDiffIndexSubset::get(
+          ctx, derivedDA->getParameterIndices()->parameters);
+      if (baseParameters->isSubsetOf(derivedParameters)) {
+        defined = true;
+        break;
+      }
+    }
+    if (!defined) {
+      // Omit printing wrt clause if attribute differentiation parameters match
+      // inferred differentiation parameters.
+      auto *inferredParameters = TypeChecker::inferDifferentiableParameters(
+          derivedAFD, nullptr);
+      bool omitWrtClause = baseDA->getParameterIndices()->parameters.count() ==
+                           inferredParameters->parameters.count();
+      // Get `@differentiable` attribute description.
+      std::string baseDAString;
+      llvm::raw_string_ostream stream(baseDAString);
+      baseDA->print(stream, derivedDecl, omitWrtClause);
+      diags.diagnose(
+          derivedDecl,
+          diag::protocol_witness_missing_differentiable_attr,
+          StringRef(stream.str()).trim());
+      return false;
+    }
+  }
+
+  // If there is no differentiable attribute in `derivedDecl`, then
+  // overriding is not allowed.
+  if (derivedDAs.empty())
+    return false;
+
+  // Finally, go through all differentiable attributes in
+  // `derivedDecl` and check if they subsume any of the
+  // differentiable attributes in `baseDecl`.
+  for (auto derivedDA : derivedDAs) {
+    auto derivedParameters = AutoDiffIndexSubset::get(
+        ctx, derivedDA->getParameterIndices()->parameters);
```

Comment:
`derivedDA->getParameterIndices()` here returns `nullptr` in some external code I am compiling. It's hard to add a test that reproduces this because the error only pops up when I merge two modules into one (the first used to contain the protocol definitions and the second had a function defined that was differentiable and invoked one of the protocol functions). I haven't been able to reproduce the error in a test but I have a separate library that currently crashes at this line.

Reply:
It would be nice if we added an assertion in `getParameterIndices()` maybe and then traced down everywhere where this happens because I believe it should never be `nullptr` when we end up using it.

Reply:
Aha, that is exactly the same serialization issue. We should omit cases where `getParameterIndices()` returns `nullptr`.

Reply:
I think this is a special thing in deserializer where declarations are processed twice.

Reply:
Interesting. I pushed a fix that omits those when iterating through the attributes. However, do you think it might be better to fix the root issue in the deserializer and avoid having them processed twice? Or would that be too cumbersome?

Reply:
It would be great to fix the root issue, though I probably don't have time to look into it in the near term.

Reply:
Could you give me some pointers? I can take a look and if it's not super time consuming for me I might tackle it.

Reply:
@dan-zheng do you remember the deserialization issue?